### PR TITLE
[PyCDE] Add new API to import existing HW modules into a System.

### PR DESCRIPTION
**Pull request by mikeurbach**
This can be used to load HW modules from an MLIR file on disk, and
insert them into a fresh System.

### CODE REVIEW

**Comment by mikeurbach** [comment]
Opening as a draft to get the ball rolling. I am going to update the System API a bit, but the change in module.py shouldn't really be changing.

**Comment by mikeurbach** [comment]
@teqdruid I think this is ready for review. PTAL whenever you get a chance.

File: frontends/PyCDE/src/pycde/module.py [hunk 1]
```diff
@@ -394,6 +394,31 @@ def externmodule(to_be_wrapped, extern_name=None):
...
+  for output_name, output_type in hw_module.outputs().items():
+    ports[output_name] = Output(output_type, output_name)
+
+  # Use the name and ports to construct a class object like what externmodule
+  # would wrap.
+  cls = type(name, (object,), ports)
+
+  # Creation callback that just moves the already build module into the System's
+  # ModuleOp and returns it.
+  def create_cb(sys: System, mod: _SpecializedModule, symbol: str):
```
Comment (teqdruid) [inline frontends/PyCDE/src/pycde/module.py:414]: I _think_ you need to store `hw_module` as a default parameter to this function so it doesn't change with the next import_hw_module call. Maybe that's only for lambdas. Python closures suck.

File: frontends/PyCDE/src/pycde/system.py [hunk 2]
```diff
@@ -89,6 +90,20 @@ def _get_ip(self):
   def set_debug():
     ir._GlobalDebug.flag = True
 
+  def import_hw_modules(self, path: Path):
```
Comment (teqdruid) [inline frontends/PyCDE/src/pycde/system.py:93]: Why muck around with file I/O? Just accept a string then require the user to load from disk. More flexibility.

File: frontends/PyCDE/test/test_import_hw_modules.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  hw.output %0 : i1
+}
+
+hw.module @and(%a: i1, %b: i1) -> (out: i1) {
+  %0 = comb.and %a, %b : i1
+  hw.output %0 : i1
+}
+""")
+tmp.flush()
+
```
Comment (teqdruid) [inline frontends/PyCDE/test/test_import_hw_modules.py:21]: Can you add a PyCDE module interacting with a loaded one? Like instantiating one? I think you need to load the symbol into a namespace of some sort.

File: frontends/PyCDE/src/pycde/module.py (hunk 1 above)
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/module.py:414]: though based on your test below, perhaps not... I don't understand Python closures.

**Review by teqdruid — APPROVED** [review]
This is _really_ slick. Thanks!

File: frontends/PyCDE/test/test_import_hw_modules.py (hunk 3 above)
Comment (mikeurbach) [reply frontends/PyCDE/test/test_import_hw_modules.py:21]: This is what i'm working on now, and will include an update in this PR. I'm realizing that we'd want to use this a little differently, and that will probably be a change in the System API in this PR based on that usage.

File: frontends/PyCDE/src/pycde/system.py (hunk 2 above)
Comment (mikeurbach) [reply frontends/PyCDE/src/pycde/system.py:93]: I think actually the API should be just a list of modules returned from `module.import_hw_module`. This seems more orthogonal, and if you are going to actually instantiate the imported modules, you'll need to import them first before declaring modules and instantiating the system, and can just pass them to this API.

File: frontends/PyCDE/src/pycde/module.py (hunk 1 above)
Comment (mikeurbach) [reply frontends/PyCDE/src/pycde/module.py:414]: yeah i think that weirdness is only for lambas.

**Comment by mikeurbach** [comment]
Ok, I pushed up a change to show how you can instantiate modules you've imported. Does this usage model make sense? I simplified the System API to reflect the fact that if you want to instantiate the modules, you will already have a list of them around to pass to the System.

File: frontends/PyCDE/test/test_import_hw_modules.py (hunk 3 above)
Comment (teqdruid) [reply frontends/PyCDE/test/test_import_hw_modules.py:24]: Yeah, this works as a base layer. I suspect we're gonna want some convenience functions on top for common use cases.

**Comment by teqdruid** [comment]
I think we'll probably have to refine the usage model over time. You're good to merge this!

**Pull request merged**